### PR TITLE
Add NullableDocument<> with deserialization support

### PR DESCRIPTION
**Pull request by pnwpedro**

### CODE REVIEW

File: Fauna/Types/NullableDocument.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+namespace Fauna.Types;
+
+/// <summary>
+/// A wrapper class that allows <see cref="Document"/>, <see cref="DocumentRef"/>, and user-defined classes
+/// to be null references.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public class NullableDocument<T>
```
Comment (pnwpedro) [inline Fauna/Types/NullableDocument.cs:8]: Consider NullableDocument abstract class with concrete NullDocument and NonNullDocument

File: Fauna/Serialization/ClassDeserializer.cs [hunk 2]
```diff
@@ -8,7 +10,7 @@ internal interface IClassDeserializer : IDeserializer
     public object DeserializeDocument(MappingContext context, string? id, string? name, ref Utf8FaunaReader reader);
 }
 
-internal class ClassDeserializer<T> : BaseDeserializer<T>, IClassDeserializer
+internal class ClassDeserializer<T> : BaseDeserializer<T>, IClassDeserializer where T : class
```
Comment (freels) [inline Fauna/Serialization/ClassDeserializer.cs:13]: is the constraint here necessary?

File: Fauna/Serialization/DocumentDeserializer.cs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            }
+
+            return (new NonNullDocument<Document>(doc) as T)!;
+        }
+
+        if (name != null && coll != null && ts != null)
+        {
+            // If we're here, name is a string.
+            var nameAsString = (string)name;
+            var r = new NamedDocument((string)name, coll, ts.GetValueOrDefault(), data);
```
Comment (freels) [inline Fauna/Serialization/DocumentDeserializer.cs:101]: use `nameAsString` here?

**Review by freels — APPROVED** [review]

File: Fauna/Serialization/ClassDeserializer.cs (hunk 2 above)
Comment (pnwpedro) [reply Fauna/Serialization/ClassDeserializer.cs:13]: Nope, this is an artifact of an impl I explored. Removing

**Pull request merged**